Allow for returning roomId in onAliasQuery

MadLittleMods opened the pull request:
Allow for returning `roomId` in `onAliasQuery`.

Instead of needing to return `creationOpts`, you can now create the room yourself and return the new `roomId` you made.

---

MR for context where we want this feature: https://gitlab.com/gitterHQ/webapp/-/merge_requests/2088

As discussed in https://matrix.to/#/!ViClwbclMrgQwyPuvH:matrix.org/$VgYNMvbNXI8wjhwD5y9kI_vuQbgWgfF1-akf2VnAaNQ?via=half-shot.uk&via=matrix.org&via=tchncs.de


## Dev notes

```
npm run test -- spec/integ/bridge.spec.js
```

MadLittleMods committed:
Allow for returning roomId in onAliasQuery

Instead of needing to return `creationOpts`, you can now create the room yourself
and return the new `roomId` you made.

MR for context where we want this feature: https://gitlab.com/gitterHQ/webapp/-/merge_requests/2088

MadLittleMods commented on spec/integ/bridge.spec.js:
```diff
@@ -187,6 +187,22 @@ describe("Bridge", function() {
+            expect(clients["bot"].createRoom).not.toHaveBeenCalled();
+
+            const room = await bridge.getRoomStore().getMatrixRoom("!abc123:bar");
+            expect(room).toBeDefined();
```
This bottom part to assert the `room` is just taken from the tests below but I am pretty confused about it testing anything.

If you actually debug this, `room` is `null` and the test seems to support both cases so that is why it passes. Seems to be a problem with all of the tests here.

Half-Shot replied:
Feels like you need to await `onAliasQuery` to ensure that the promises all run here? This might be a bug with the other tests.

Half-Shot replied:
With:
```js
            bridgeCtrl.onAliasQuery.and.returnValue({ roomId: "!abc123:bar" });
            await bridge.run(101, {}, appService);

            await appService.onAliasQuery("#foo:bar");

            expect(clients["bot"].createRoom).not.toHaveBeenCalled();

            const room = await bridge.getRoomStore().getMatrixRoom("!abc123:bar");
            expect(room).toBeDefined();
            expect(room.getId()).toEqual("!abc123:bar");
```

it seems to pass :)

MadLittleMods replied:
Thanks for spotting that! Just copy pasting everything from the other tests. We should fix those up too in another PR

MadLittleMods replied:
Created https://github.com/matrix-org/matrix-appservice-bridge/pull/289 to fix the other tests

MadLittleMods commented on spec/integ/bridge.spec.js:
```diff
@@ -187,6 +187,22 @@ describe("Bridge", function() {
+            expect(clients["bot"].createRoom).not.toHaveBeenCalled();
+
+            const room = await bridge.getRoomStore().getMatrixRoom("!abc123:bar");
+            expect(room).toBeDefined();
```
This bottom part to assert the `room` is just taken from the tests below but I am pretty confused about it testing anything.

If you actually debug this, `room` is `null` and the test seems to support both cases so that is why it passes. Seems to be a problem with all of the tests here.

MadLittleMods commented on spec/integ/bridge.spec.js:
```diff
@@ -187,6 +187,22 @@ describe("Bridge", function() {
             });
```
I am seeing linting and test errors for something I didn't change. Am I doing something wrong?

https://buildkite.com/matrix-dot-org/matrix-appservice-bridge/builds/411#dad87f0f-d7c1-4a2c-a9e6-8e9ea816d86d
```
src/components/client-factory.ts:128:62 - error TS2731: Implicit conversion of a 'symbol' to a 'string' will fail at runtime. Consider wrapping this expression in 'String(...)'.
--
  |  
  | 128                 const loggerInstance = logging.get(`js-sdk:${loggerName}`);
```

Half-Shot replied:
This is https://github.com/matrix-org/matrix-appservice-bridge/pull/286

Half-Shot commented on src/bridge.ts:
```diff
@@ -85,7 +85,7 @@ export interface BridgeController {
      * will automatically be stored in the associated `roomStore`. */
     onAliasQuery?: (alias: string, aliasLocalpart: string) =>
-        PossiblePromise<{creationOpts: Record<string, unknown>, remote?: RemoteRoom}|null|void>;
+        PossiblePromise<{roomId?: string, creationOpts: Record<string, unknown>, remote?: RemoteRoom}|null|void>;
```
You are no longer required to pass these in:

```suggestion
        PossiblePromise<{roomId?: string, creationOpts?: Record<string, unknown>, remote?: RemoteRoom}|null|void>;
```

MadLittleMods replied:
Great spot! :+1:

Half-Shot commented on spec/integ/bridge.spec.js:
```diff
@@ -187,6 +187,22 @@ describe("Bridge", function() {
         });
 
+        it("should not create a room if roomId is returned from the function but should still store it",
+        async function(done) {
```
I don't think you need `done` here. Some tests use it because they were written long before `async` was a thing.

Half-Shot commented on spec/integ/bridge.spec.js:
```diff
@@ -187,6 +187,22 @@ describe("Bridge", function() {
+
+            const room = await bridge.getRoomStore().getMatrixRoom("!abc123:bar");
+            expect(room).toBeDefined();
+            if (!room) { done(); return; }
```
I'm concerned about this, if you await onAliasQuery you shouldn't need this?

MadLittleMods committed:
Fix runtime conversion that typescript complains about

Add https://github.com/matrix-org/matrix-appservice-bridge/pull/286 to my PR so the CI passes
